### PR TITLE
undocument curl (Mai 2018 removal date)

### PR DESCRIPTION
**Pull request by burner**
* undocument mai 2016
* deprecation mai 2017
* removal mai 2018

### CODE REVIEW

File: etc/c/curl.d [hunk 1]
```diff
@@ -1,4 +1,7 @@
 /**
+   $(RED Deprecated: This module is considered out-dated and not up to Phobos'
+   current standards. It will be remove in Mai 2018.)
```
Comment (UplinkCoder) [inline etc/c/curl.d]: "remove in Mai 2018" -> "removed in May 2018"

Comment (Philpax) [inline etc/c/curl.d]: "remove in Mai 2018" -> "removed in May 2018"

File: std/net/curl.d [hunk 2]
```diff
@@ -1,6 +1,9 @@
 // Written in the D programming language.
 
 /**
+$(RED Deprecated: This module is considered out-dated and not up to Phobos'
+current standards. It will be remove in Mai 2018.)
```
Comment (Philpax) [inline std/net/curl.d]: ditto

File: etc/c/curl.d (hunk 1 above)
Comment (burner) [inline etc/c/curl.d]: done

File: std/net/curl.d (hunk 2 above)
Comment (burner) [inline std/net/curl.d]: done

Comment (burner) [inline std/net/curl.d]: thanks

**Comment by Philpax** [comment]
"remove" still needs to be changed to "removed" - otherwise LGTM

**Comment by burner** [comment]
also done

**Comment by JackStouffer** [comment]
Hold the phone, when was this decided? The curl binding's poor quality was talked about during the panel today but don't you think removal deserves at least a news group post?

**Comment by 9il** [comment]
btw, what can be used without vibe.d instead? Do we plan to add something instead of net.curl?

**Comment by ZombineDev** [comment]
BTW, after deprecation (if not sooner) curl and friends should be moved to unDead or somewhere else on code.dlang.org, to lessen the breakage.

**Comment by burner** [comment]
the deprecation path by now seams to be.

1. remove doc
2. mark as deprecated
3. delete in phobos and move to undead

@DmitryOlshansky there will be now replacement for curl in phobos.

**Comment by burner** [comment]
http://forum.dlang.org/post/nhquzxmhwvawlxatzytr@forum.dlang.org

**Comment by wilzbach** [comment]
wasn't it decided that __entire `etc.c`__ must go? So `zlib` and `sqlite3` too?

**Comment by burner** [comment]
@wilzbach yes, I will create these PRs next!

**Comment by schveiguy** [comment]
2 recommendations:

1. Do not remove curl docs. If you are deprecating the module, but we are including it, I think we should document that it's there and deprecated.
2. No reason to put "This isn't up to Phobos standards". Just put that it's deprecated and when this will happen.

**Comment by CyberShadow** [comment]
> @DmitryOlshansky there will be now replacement for curl in phobos.

What is "now"? "no"? "new"?

Currently, curl is the only reasonable way to do a HTTP request with D without third-party libraries.

**Comment by 9il** [comment]
> @DmitryOlshansky there will be now replacement for curl in phobos.
> What is "now"? "no"? "new"?
> 
> Currently, curl is the only reasonable way to do a HTTP request with D without third-party libraries.
>  Show all checks

So, I suppose "@DmitryOlshansky" is "@9il" and "now" is "new" =)

**Comment by burner** [comment]
IMO it is not reasonable, it is blocking

**Comment by CyberShadow** [comment]
> So, I suppose "@DmitryOlshansky" is "@9il" and "now" is "new" =)

What's the new replacement?

**Comment by CyberShadow** [comment]
> IMO it is not reasonable, it is blocking

It's OK for many scripts and small programs...

**Comment by schveiguy** [comment]
@CyberShadow curl will still be there, the bindings will exist in undead.

Curl is already a 3rd party library (not always installed), this is just a binding. It's definitely good work, and should remain, but adds external dependencies that simply shouldn't be part of the core language.

**Comment by CyberShadow** [comment]
I think that until we get a native HTTP library in Phobos, this goes against the "first 5 minutes" idea. Downloading a web page or doing a REST request is one of the things people would want to do in small D programs without any dependencies. Telling people that they have to get another library or write a `dub.sdl` file for such a simple thing would not be so great.

**Comment by burner** [comment]
curl, sqlite and everything binding directly to not OS C stuff is just a pain to maintain. It should just go.

**Comment by schveiguy** [comment]
> this goes against the "first 5 minutes" idea.

Seriously, if the first 5 minutes is "learn how to use curl", then we aren't doing much better.

**Comment by andralex** [comment]
We need a replacement before we deprecate curl.

**Comment by CyberShadow** [comment]
> Seriously, if the first 5 minutes is "learn how to use curl", then we aren't doing much better.

It's not that bad. `std.net.curl` does a very nice job of wrapping everything in a curl-agnostic manner. Really, just from looking at its public API, it's actually not easy to tell that it's a curl wrapper.

Speaking of which, that creates a deprecation path - it might be possible to create an API implementation for std.net.curl which underneath uses native D code in the future.

**Comment by burner** [comment]
@CyberShadow that is properly the way to go.

**Comment by wilzbach** [comment]
> We need a replacement before we deprecate curl.

Just a thought from my (newcomer) view - we should document that this on our agenda!
Maybe we can have a (high-level) community roadmap?

I updated our wish list recently, but it still feels a bit out of date.
https://wiki.dlang.org/Wish_list
https://wiki.dlang.org/How_You_Can_Help

Better ideas? Maybe we can find a better place or format?

**Comment by schveiguy** [comment]
@CyberShadow apologies, I was mistakenly looking at etc.c.curl as the only module we had for curl in D, which didn't look that user-friendly.

I agree if we can replace std.net.curl with something that uses D underneath is likely useful. I'm just a bit skeptical that someone will produce an equivalent to curl for D any time soon.

In any case, I obviously am not qualified to comment further, as I'm not a user of this feature.

**Comment by GilesBathgate** [comment]
@burner OK I have made a start on the replacement library for __std.net.curl__, I am having some difficulties though perhaps you can help me out.

```d
extern(C) int tcp_connect(void* socket);

void main()
{
  static immutable string paula = "Brillant";
  tcp_connect(cast(void*)&paula);
}
```

I am completely kidding of course!

**Comment by GilesBathgate** [comment]
>static immutable string paula = "Brillant"; 

@burner That probably wasn't funny unless you were familiar with this reference: http://thedailywtf.com/articles/The_Brillant_Paula_Bean

**Comment by sigod** [comment]
> That probably wasn't funny unless you were familiar with this reference: http://thedailywtf.com/articles/The_Brillant_Paula_Bean

I'd say it's too funny.
screenshot 2016-05-17 22 06 19

**Pull request closed without merging**